README.md
# Truly infinite list

This package maybe is the Most *Simplest* Wrap for truly infinite list.

## Features:

- Fixed memory consumption, only cache fixed number of items in mempory.
- Progressive placeholder for async loading.
- An async fetching function for a page of items.
- Support ListView, GridView and SliverList.
 
## Screenshot

When Pull up, the progressive placeholder indicate loading.

![loading placeholder](https://github.com/pedia/infinite_pagination/raw/master/images/pull-up.png)

When Drag down, out of screen items was evicted is loading again.

![loading placeholder](https://github.com/pedia/infinite_pagination/raw/master/images/drag-down.png)

Animation demo.

![gif](https://github.com/pedia/infinite_pagination/raw/master/images/full-demo.gif)


## How to use

To use this plugin, add infinite_pagination as a dependency in your pubspec.yaml file. For example:

```
dependencies:
  infinite_pagination:
  provider:
```

Usually we use [provider](https://pub.dev/packages/provider) as State Management.


```dart
import 'package:infinite_pagination/infinite_pagination.dart';


class Catalog extends InfinitePagination<Chapter> {
  Catalog()
      : super(
          fetcher: (int startingIndex, int itemsPerPage) async => 
              fetchPage<Chapter>(
            startingIndex: startingIndex,
            countPerPage: 10,
            total: 142,
            mock: (i) => Chapter(i),
            delay: 500,
          ),
        );
}

Widget build(BuildContext context) {
    return ChangeNotifierProvider(
      create: (_) => Catalog(),
      child: Selector<Catalog, int>(
          selector: (context, catalog) => catalog.itemCount,
          builder: (context, itemCount, child) => ListView.builder(
            key: PageStorageKey('ListView'),
            itemCount: itemCount,
            itemBuilder: (context, index) {
              // Every item of the `ListView` is individually listening
              // to the catalog.
              var catalog = Provider.of<Catalog>(context);

              // Catalog provides a single synchronous method for getting
              // the current data.
              var item = catalog.getByIndex(index);

              if (item == null) {
                return LoadingTile();
              }

              return ListTile(
                leading: CircleAvatar(child: Text('${item.i}')),
                title: Text('Chapter #${item.i}'),
                subtitle: Text('subtitle ${item.i}'),
              );
            },
          ),
        ),
    );
}
```

## Best experience of saving memory

- Use [file_cache](https://pub.dev/packages/file_cache) cache Http Request in file system.
- Use this package indicate loading and infinite paging.


This package is inspired by(copy from) [official sample](https://github.com/flutter/samples/tree/master/infinite_list) 


infinite_pagination/CHANGELOG.md
## [0.0.2] - Fixed link

* Fixed pub score.

## [0.0.1] - First initialized version

* Add InfinitePagination, Pagination, LoadingTile and fetchPage


test/infinite_pagination_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:infinite_pagination/infinite_pagination.dart';

class Pool extends InfinitePagination<int> {
  Pool({this.values});

  final List<int> values;

  Future<Pagination<int>> fetch(int startingIndex, int itemsPerPage) async {
    final results = values.length > startingIndex
        ? values.sublist(startingIndex, startingIndex + itemsPerPage)
        : <int>[];

    return Pagination<int>(
      items: results,
      startingIndex: startingIndex,
      hasNext: values.length > startingIndex + itemsPerPage,
    );
  }
}

void main() {
  test('init state', () {
    final pool = InfinitePagination<int>();
    expect(pool.fetcher, isNull);
    expect(pool.itemsPerPage, equals(10));
    expect(pool.itemCount, isNull);

    // TODO:
    // expect(pool.getByIndex(0), isNull);
    // expect(pool.getByIndex(1), throwsException);
  });

  testWidgets('normal pool', (WidgetTester tester) async {
    final pool = Pool(values: List.generate(200, (i) => i));
    expect(pool.fetcher, isNull);
    expect(pool.itemsPerPage, equals(10));
    expect(pool.itemCount, isNull);

    expect(pool.getByIndex(0), isNull);
    await tester.pump(Duration.zero);
    expect(pool.getByIndex(0), 0);
    expect(pool.itemCount, isNull);

    expect(pool.pages.length, equals(1));
    expect(
        pool.toString(),
        equals(
            'InfinitePagination(total: null page: 1/(0) fetched page: {}/{})'));

    expect(pool.getByIndex(1), 1);
    expect(pool.getByIndex(9), 9);
    expect(pool.getByIndex(10), isNull);

    for (int page = 1; page < 200 ~/ 10; page++) {
      expect(pool.getByIndex(page * pool.itemsPerPage), isNull);
      await tester.pump(Duration.zero);
      expect(
          pool.getByIndex(page * pool.itemsPerPage), page * pool.itemsPerPage);

      for (int n = page * pool.itemsPerPage;
          n < page * pool.itemsPerPage + pool.itemsPerPage;
          n++) expect(pool.getByIndex(n), n);
    }
  });

  testWidgets('empty pool', (tester) async {
    final pool = Pool(values: <int>[]);

    final page = await pool.fetch(0, 10);
    expect(page.items.isEmpty, isTrue);

    expect(pool.fetcher, isNull);
    expect(pool.itemsPerPage, equals(10));
    expect(pool.itemCount, isNull);

    expect(pool.getByIndex(0), isNull);
    await tester.pump(Duration.zero);
    expect(pool.getByIndex(0), isNull);

    expect(pool.itemCount, 0);
    expect(pool.toString(),
        equals('InfinitePagination(total: 0 page: 1/(0) fetched page: {}/{})'));
    expect(pool.pages[0].hasNext, isFalse);
    expect(pool.itemCount, 0);

    expect(pool.getByIndex(1), isNull);
    expect(pool.getByIndex(2), isNull);
    expect(pool.itemCount, equals(0));

    expect(pool.getByIndex(10), isNull);
    await tester.pump(Duration.zero);
    expect(pool.getByIndex(10), isNull);
    expect(pool.toString(),
        equals('InfinitePagination(total: 0 page: 1/(0) fetched page: {}/{})'));
    expect(pool.getByIndex(11), isNull);
    expect(pool.getByIndex(100), isNull);
    expect(pool.getByIndex(200), isNull);
    expect(pool.itemCount, equals(0));
  });
}


lib/nasa_demo.dart
// import 'dart:io';

import 'package:flutter/material.dart';
import 'package:provider/provider.dart';

import 'package:infinite_pagination/infinite_pagination.dart';
import 'package:file_cache/file_cache.dart';

void main() {
  // In Nasa image Http Header:
  //    cache-control: public, max-age=300, s-max-age=600
  // This is too short, force to cache 10 days.
  FileCache.forceCacheSeconds = 86400 * 100;

  runApp(NasaDemoApp());
}

class Link {
  Link({this.url, this.rel, this.render});

  final String url;
  final String rel;
  final String render;
}

class NasaMedia {
  NasaMedia({
    this.url,
    this.creator,
    this.description,
    this.id,
    this.created,
    this.keywords,
    this.title,
    this.center,
    this.mediaType,
    this.links,
  });

  final String url;
  final String creator;
  final String description;
  final String id;
  final DateTime created;
  final List<String> keywords;
  final String title;
  final String center;
  final String mediaType;
  final List<Link> links;

  factory NasaMedia.fromJson(Map map) {
    if (map is Map) {
      final data = map['data'].first as Map;

      return NasaMedia(
        url: map['href'],
        creator: data['creator'],
        description: data['description'],
        id: data['nasa_id'],
        created: DateTime.parse(data['date_created']),
        keywords: (data['keywords'] as List).cast<String>(),
        title: data['title'],
        center: data['center'],
        mediaType: data['media_type'],
        links: (map['links'] as List).map((i) {
          return Link(
            url: i['href'],
            rel: i['preview'],
            render: i['image'],
          );
        }).toList(),
      );
    } else
      throw ArgumentError('Map must be Map, current is ${map.runtimeType}');
  }
}

class NasaLibrary extends InfinitePagination<NasaMedia> {
  NasaLibrary() : super(itemsPerPage: 100, maxCacheDistance: 200);

  final Future<FileCache> fileCache = FileCache.fromDefault();

  fetch(int startingIndex, int itemsPerPage) async {
    // Since Nasa does not support this argument, fixed as 100
    itemsPerPage = 100;

    String url = startingIndex == 0
        ? 'https://images-api.nasa.gov/search?q=earth&media_type=image'
        : 'https://images-api.nasa.gov/search?q=earth&media_type=image&page=${startingIndex ~/ itemsPerPage}';

    Map jd = await (await fileCache).getJson(url);

    final items = (jd['collection']['items'] as List)
        .map((i) => NasaMedia.fromJson(i))
        .toList();

    return Pagination(
      items: items,
      startingIndex: startingIndex,
      hasNext: items.length == itemsPerPage,
    );
  }
}

class NasaDemoApp extends StatefulWidget {
  @override
  _DemoAppState createState() => _DemoAppState();
}

class _DemoAppState extends State<NasaDemoApp>
    with SingleTickerProviderStateMixin {
  TabController _tabController;

  @override
  void initState() {
    _tabController = TabController(length: 2, vsync: this);
    super.initState();
  }

  @override
  void dispose() {
    _tabController.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return ChangeNotifierProvider(
      create: (_) => NasaLibrary(),
      child: MaterialApp(
        debugShowCheckedModeBanner: false,
        theme: ThemeData(
          primarySwatch: Colors.blue,
          visualDensity: VisualDensity.adaptivePlatformDensity,
        ),
        home: Material(
          child: Selector<NasaLibrary, int>(
            selector: (context, catalog) => catalog.itemCount,
            builder: (context, itemCount, child) => ListView.builder(
              key: PageStorageKey('ListView'),
              itemCount: itemCount,
              itemBuilder: (context, index) {
                // Every item of the `ListView` is individually listening
                // to the catalog.
                var catalog = Provider.of<NasaLibrary>(context);

                // Catalog provides a single synchronous method for getting
                // the current data.
                var item = catalog.getByIndex(index);

                if (item == null) {
                  return LoadingTile();
                }

                return Card(
                    color: Colors.black,
                    child: Stack(
                      children: [
                        Center(
                          child: Image(
                            image: FileCacheImage(item.links[0].url),
                            fit: BoxFit.fitWidth,
                          ),
                        ),
                        Positioned(
                          left: 8,
                          right: 8,
                          top: 8,
                          child: Text(
                            item.description,
                            style: TextStyle(color: Colors.white),
                            maxLines: 2,
                            overflow: TextOverflow.ellipsis,
                          ),
                        ),
                        Positioned(
                          left: 8,
                          right: 8,
                          bottom: 8,
                          child: Text(
                            '${item.title} | ${item.created}',
                            style: TextStyle(color: Colors.white),
                            maxLines: 1,
                            overflow: TextOverflow.ellipsis,
                          ),
                        ),
                      ],
                    ));
              },
            ),
          ),
        ),
      ),
    );
  }
}


lib/mock_fetch.dart
import 'dart:math';
import 'infinite_pagination.dart';

/// This function emulates a REST API call. You can imagine replacing its
/// contents with an actual network call, keeping the signature the same.
///
/// It will fetch a page of items from [startingIndex].
Future<Pagination<T>> fetchPage<T>({
  int startingIndex,
  int total = 100,
  int countPerPage = 10,
  T mock(int index),
  int delay = 300,
}) async {
  // We're emulating the delay inherent to making a network call.
  await Future<void>.delayed(Duration(milliseconds: delay));

  // If the [startingIndex] is beyond the bounds of the catalog, an
  // empty page will be returned.
  if (startingIndex > total) {
    return Pagination<T>(
      items: <T>[],
      startingIndex: startingIndex,
      hasNext: false,
    );
  }

  int left = min(total - startingIndex, countPerPage);

  // The page of items is generated here.
  return Pagination<T>(
    items: List<T>.generate(left, (index) => mock(startingIndex + index)),
    startingIndex: startingIndex,
    // Returns `false` if we've reached the [total].
    hasNext: startingIndex + countPerPage < total,
  );
}


lib/infinite_pagination.dart
library infinite_pagination;

import 'package:flutter/material.dart';

/// Contained one page of items
class Pagination<T> {
  Pagination({
    this.items,
    this.startingIndex = 0,
    this.hasNext = false,
  });

  final List<T> items;
  final int startingIndex;
  final bool hasNext;
}

class InfinitePagination<T> extends ChangeNotifier {
  InfinitePagination({
    this.fetcher,
    this.itemsPerPage = 10,
    this.maxCacheDistance = 100,
  }) : assert(maxCacheDistance > itemsPerPage);

  /// Async function for fetching a page of items, usually this is an Http
  /// request.
  ///
  /// ```dart
  /// Future <Pagination<T>> fetchPage(int startingIndex, int itemsPerPage) async
  /// ```
  /// If [fetcher] is provide, we should extend from [InfinitePagination] and
  /// implement [fetch].
  final Future<Pagination<T>> Function(int, int) fetcher;

  ///
  final int itemsPerPage;

  /// This is the maximum number of the items we want in memory in each
  /// direction from the current position. For example, if the user
  /// is currently looking at item number 400, we don't want item number
  /// 0 to be kept in memory.
  /// [maxCacheDistance] should bigger than item count in a screen
  final int maxCacheDistance;

  /// The internal store of pages that we got from [fetchPage].
  /// The key of the map is the starting index of the page, for faster
  /// access.
  final Map<int, Pagination<T>> _pages = {};

  @visibleForTesting
  get pages => _pages;

  /// A set of pages (represented by their starting index) that have started
  /// the fetch process but haven't ended it yet.
  ///
  /// This is to prevent fetching of a page several times in a row. When a page
  /// is already being fetched, we don't initiate another fetch request.
  final Set<int> _pagesBeingFetched = {};

  /// Total size of items. This is `null` at first, and only when the user
  /// reaches the end, it will hold the actual number.
  int itemCount;

  /// After disposed, we don't allow it to call
  /// [notifyListeners].
  bool _isDisposed = false;

  @override
  void dispose() {
    _isDisposed = true;
    super.dispose();
  }

  Future<Pagination<T>> fetch(int startingIndex, int itemsPerPage) async {
    assert(fetcher != null);
    return fetcher(startingIndex, itemsPerPage);
  }

  /// This is a synchronous method that returns the item at [index].
  ///
  /// If the item is already in memory, this will just return it. Otherwise,
  /// this method will initiate a fetch of the corresponding page, and will
  /// return [Item.loading].
  ///
  /// The UI will be notified via [notifyListeners] when the fetch
  /// is completed. At that time, calling this method will return the newly
  /// fetched item.
  T getByIndex(int index) {
    // Compute the starting index of the page where this item is located.
    // For example, if [index] is `42` and [itemsPerPage] is `20`,
    // then `index ~/ itemsPerPage` (integer division)
    // evaluates to `2`, and `2 * 20` is `40`.
    var startingIndex = (index ~/ itemsPerPage) * itemsPerPage;

    // If the corresponding page is already in memory, return immediately.
    if (_pages.containsKey(startingIndex)) {
      if (index - startingIndex < _pages[startingIndex].items.length) {
        var item = _pages[startingIndex].items[index - startingIndex];
        return item;
      }
    }

    // We don't have the data yet. Start fetching it.
    if (itemCount == null) {
      _fetchPage(startingIndex).then((_) {});
    }

    // In the meantime, return a placeholder.
    return null;
  }

  /// This method initiates fetching of the [ItemPage] at [startingIndex].
  Future<void> _fetchPage(int startingIndex) async {
    if (_pagesBeingFetched.contains(startingIndex)) {
      // Page is already being fetched. Ignore the redundant call.
      return;
    }

    _pagesBeingFetched.add(startingIndex);
    final page = await fetch(startingIndex, itemsPerPage);
    _pagesBeingFetched.remove(startingIndex);

    if (!page.hasNext && itemCount == null) {
      // The returned page has no next page. This means we now know the size.
      itemCount = startingIndex + page.items.length;
    }

    // Store the new page.
    _pages[startingIndex] = page;
    _pruneCache(startingIndex);

    if (!_isDisposed) {
      // Notify the widgets that are listening to that they
      // should rebuild.
      notifyListeners();
    }
  }

  /// Removes item pages that are too far away from [currentStartingIndex].
  void _pruneCache(int currentStartingIndex) {
    // It's bad practice to modify collections while iterating over them.
    // So instead, we'll store the keys to remove in a separate Set.
    final keysToRemove = <int>{};
    for (final key in _pages.keys) {
      if ((key - currentStartingIndex).abs() > maxCacheDistance) {
        // This page's starting index is too far away from the current one.
        // We'll remove it.
        keysToRemove.add(key);
      }
    }
    for (final key in keysToRemove) {
      _pages.remove(key);
    }
  }

  bool get isEmpty => _pages.isNotEmpty && itemCount == 0;

  @override
  String toString() {
    return 'InfinitePagination(total: $itemCount page: ${_pages.length}/${_pages.keys} '
        'fetched page: $_pagesBeingFetched/${_pagesBeingFetched})';
  }

  void clear() {
    _pages.clear();
    _pagesBeingFetched.clear();
    itemCount = null;

    if (!_isDisposed) notifyListeners();
  }
}

/// This is the widget responsible for building the "still loading" item
/// in the list (represented with "..." and a crossed square).
class LoadingTile extends StatelessWidget {
  const LoadingTile({Key key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    final isDark = Theme.of(context).backgroundColor.computeLuminance() < 0.179;

    final color = isDark ? Colors.grey[850] : Colors.grey[200];

    return Padding(
      padding: const EdgeInsets.all(8.0),
      child: ListTile(
        leading: AspectRatio(
          aspectRatio: 1,
          child: Container(color: color),
        ),
        title: Container(
          color: color,
          margin: EdgeInsets.only(right: 100),
          height: 24,
        ),
        subtitle: Container(
          color: color,
          height: 14,
        ),
      ),
    );
  }
}


lib/main.dart
import 'package:flutter/material.dart';
import 'package:provider/provider.dart';

import 'package:infinite_pagination/infinite_pagination.dart';
import 'package:infinite_pagination/mock_fetch.dart';

void main() {
  runApp(DemoApp());
}

/// Fake item for Demo
class Chapter {
  Chapter(this.i);

  final int i;
}

/// Catelog container some chapters.
///
/// For simply lately used, declare a class instead use [InfinitePagination] directly.
class Catalog extends InfinitePagination<Chapter> {
  Catalog()
      : super(
          fetcher: (int startingIndex, int itemsPerPage) async =>

              // fake fetching
              fetchPage<Chapter>(
            startingIndex: startingIndex,
            countPerPage: 10,
            total: 142,
            mock: (i) => Chapter(i),
            delay: 500,
          ),
        );
}

class DemoApp extends StatefulWidget {
  @override
  _DemoAppState createState() => _DemoAppState();
}

class _DemoAppState extends State<DemoApp> with SingleTickerProviderStateMixin {
  TabController _tabController;

  @override
  void initState() {
    _tabController = TabController(length: 2, vsync: this);
    super.initState();
  }

  @override
  void dispose() {
    _tabController.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return ChangeNotifierProvider(
      create: (_) => Catalog(),
      child: MaterialApp(
        debugShowCheckedModeBanner: false,
        theme: ThemeData(
          primarySwatch: Colors.blue,
          visualDensity: VisualDensity.adaptivePlatformDensity,
        ),
        home: Material(
          child: Scaffold(
              appBar: AppBar(
                title: Text('Truly Infinite List Demo'),
                bottom: TabBar(
                  isScrollable: true,
                  controller: _tabController,
                  tabs: [
                    Text('ListView'),
                    Text('GridView'),
                    // Text('SliverList'),
                  ],
                ),
                elevation: 0,
              ),
              body: TabBarView(
                controller: _tabController,
                children: [
                  Selector<Catalog, int>(
                    selector: (context, catalog) => catalog.itemCount,
                    builder: (context, itemCount, child) => ListView.builder(
                      key: PageStorageKey('ListView'),
                      itemCount: itemCount,
                      itemBuilder: (context, index) {
                        // Every item of the `ListView` is individually listening
                        // to the catalog.
                        var catalog = Provider.of<Catalog>(context);

                        // Catalog provides a single synchronous method for getting
                        // the current data.
                        var item = catalog.getByIndex(index);

                        if (item == null) {
                          return LoadingTile();
                        }

                        return ListTile(
                          leading: CircleAvatar(child: Text('${item.i}')),
                          title: Text('Chapter #${item.i}'),
                          subtitle: Text('subtitle ${item.i}'),
                        );
                      },
                    ),
                  ),
                  Selector<Catalog, int>(
                    selector: (context, catalog) => catalog.itemCount,
                    builder: (context, itemCount, child) => GridView.builder(
                      key: PageStorageKey('GridView'),
                      gridDelegate: SliverGridDelegateWithFixedCrossAxisCount(
                        crossAxisCount: 5,
                        crossAxisSpacing: 5,
                        mainAxisSpacing: 5,
                      ),
                      itemBuilder: (context, index) {
                        var catalog = Provider.of<Catalog>(context);

                        var item = catalog.getByIndex(index);

                        if (item == null) {
                          return CircleAvatar(
                            child: Text('...'),
                            backgroundColor: Colors.grey[200],
                          );
                        }

                        return CircleAvatar(child: Text('${item.i}'));
                      },
                      itemCount: itemCount,
                    ),
                  ),
                ],
              )),
        ),
      ),
    );
  }
}
